### DISCUSSION TITLE
Strategies for recovering and applying Type information to a `json_agg` response

### DISCUSSION THREAD

**Discussion by doulighan** [sqlalchemy/sqlalchemy#11797] [Usage Questions] [answered]
I find myself somewhat regularly in a situation where I use SQLA to build a query like the following:

```sql
SELECT 'records'                              AS record_type,
       json_group_array(row_to_json(records)) AS json_agg
FROM (SELECT 
          my_table.something
          # ...
      FROM my_table
      WHERE my_table.something = 'something'
      ) AS records
UNION ALL
SELECT 'stats'                              AS record_type,
       json_group_array(row_to_json(stats)) AS json_agg
FROM (SELECT 
          AVG(my_table.something) as avg,
          COUNT(CASE WHEN my_table.other > 1000000 THEN 1 END) as count
          # ...
      FROM my_table
      WHERE my_table.something = 'something'
    ) AS stats
```

Response: 

```python
result = [
  { "record_type": "records", "json_agg": '[ { "something": 4" },  { "something": 5 }, ... ]' }
  { "record_type": "stats", "json_agg": '{ "avg": 102, "count": 10, ... }'  }
]
```

Often the columns and aggregates rely on custom types or custom `@compiles` structures, and can get pretty complicated. This works great for outputting the SQL, but obviously once I `json_agg` the result into strings, SQLA has no method of applying that type information to the response. This can be pretty painful for custom types, requiring some hard-coded work arounds.

I'm wondering if theres some kind of general strategy I could use to have SQLA deserialize the result, once I convert the JSON back into pythonic datatypes. Presumably, if I traverse down my query statement to the first `AVG` element, it might have the correct type information stored somewhere inside it. Maybe I can then tell SQLA to apply that type/transformation to a particular row of the parsed JSON that shares the same label?

I don't know, not exactly sure. Is there an idea I can use here?

**Comment by CaselIT** [comment]
Hi,

I'm not following. Your query is outputting two columns: a text one and a json one that seems to be interpreted as a string. If your question is "can sqlalchemy parse the json column instead of returning a string", yes, in the final select use type_coerce to indicate that the column is a json

Something like this should work
```py
stmt = select(sa.literal('records'), type_coerce(the_json_agg_col, JSON)).union_all(the_stats_query)
```

if that's not what you mean, then sorry I didn't understand what you are looking for

**Comment by zzzeek** [comment]
why do you need to convert SQL to JSON and then back out again?   you could use the SQLAlchemy ORM and get objects from database results normally

**Reply by doulighan** [reply]
The point is speed, we see better performance by finishing this in a single statement vs a transaction. In situations where we don't need both an aggregation and a record set returned at the same time, we don't use any JSON weirdness.

I used postgres syntax here, but this is happening in an analytics context (DuckDB and starrocks), where saving processing time is worth it.

**Reply by zzzeek** [reply]
What do you want as an end result, Python objects with fields?  those aren't going to be fast to construct no matter what you're using.

if you want JSON to Python object then I'd use whatever the current framework is for that, like pydantic or whatever

**Reply by doulighan** [reply]
In the simplest case, if I had a `MyTable.something: Mapped[Decimal] = mapped_column(Double)`, it would be nice to get a Decimal back out from the JSON. Or `Mapped[DateTime]`, etc. Of course SQLA cannot do this automatically, that wouldn't make sense. But it is true that I still have the subquery ORM object, and it should know the type of each SELECT element within it. Maybe theres some way to deserialize that JSON using those types.

In more complicated cases, this would involve TypeDecorators and calling `process_result_value`.

As I'm typing I'm sort of realizing how non standard this is, and if theres no obvious answer here I totally understand

**Reply by CaselIT** [reply]
Note that if you have an analitycal db it's likely that the time needed to load the result set into orm object, that can be 10s of thousands of rows quite easily if you are using duckdb or starrocks, may be several times more than the time used by the db to return the results.

That aside, there isn't much public to transform the list of json rows you have automatically back into sqlalchemy objects. You could likely create manually a result object with the proper processors setup, but I'm not sure if it's worth it.

**Reply by doulighan** [reply]
The worries about performance are all DB side, the returned data is paginated / aggregated. But yeah, if there's no obvious way to get this done then I'll drop it.  Thanks for responding.

**Reply by zzzeek** [reply]
the claim that you can get things back faster using json_agg vs. not would be something I'd need to understand better, a demonstration would be enlightening

**Reply by doulighan** [reply]
It might be be a bad assumption that stuck around too long, I'm going to run some comparisons. Although maybe combining them opens up some query plan optimizations.

**Reply by CaselIT** [reply]
I think the main advantage here single query with union all, not really the json agg. That's just used to transform different selects in compatible ones required by the union all

**Reply by doulighan** [reply]
Yeah thats right, the JSON agg just allows disparate results to come back in the same record without too much of a headache.

From testing it does appear that the UNION is ~10-15% faster round-trip compared with emitting two statements, with 200 records of 10 columns each.  I'm guessing this is mostly network time. This is going to be extremely dependent on the situation, but in my case thats 3s to 2.5s, which seems significant enough to continue using.

**Reply by CaselIT** [reply]
your test is between
```sql
SELECT 'records'                              AS record_type,
       json_group_array(row_to_json(records)) AS json_agg
FROM (SELECT 
          my_table.something
          # ...
      FROM my_table
      WHERE my_table.something = 'something'
      ) AS records
UNION ALL
SELECT 'stats'                              AS record_type,
       json_group_array(row_to_json(stats)) AS json_agg
FROM (SELECT 
          AVG(my_table.something) as avg,
          COUNT(CASE WHEN my_table.other > 1000000 THEN 1 END) as count
          # ...
      FROM my_table
      WHERE my_table.something = 'something'
    ) AS stats
```
and the two
```sql
SELECT 'records'                              AS record_type,
       json_group_array(row_to_json(records)) AS json_agg
FROM (SELECT 
          my_table.something
          # ...
      FROM my_table
      WHERE my_table.something = 'something'
      ) AS records
```
```sql
SELECT 'stats'                              AS record_type,
       json_group_array(row_to_json(stats)) AS json_agg
FROM (SELECT 
          AVG(my_table.something) as avg,
          COUNT(CASE WHEN my_table.other > 1000000 THEN 1 END) as count
          # ...
      FROM my_table
      WHERE my_table.something = 'something'
    ) AS stats
```
?

removing the aggregation step make skew things since it's not free to fetch possible a large number of rows

**Reply by doulighan** [reply]
Yup exactly, I'm testing it against both. Its looking like the Records portion is completely negligible compared to the Aggregates, in terms of EXPLAIN ANALYZE planning/execution time. So what we're left with is 1 slow query (UNION), or 1 slow query and 1 quick query (Aggregates then Records). That extra round trip is always going to take longer.

**Reply by CaselIT** [reply, marked as answer]
my suggestion here would be:
- if you need to use the result as json (say return form the api, load a pandas dataframe or similar) go with your query structure;
- if you want to get a list of orm instances splitting the query is likely the easiest solution, at the cost of a bit more db latency.

as always, your mileage by vary
